### PR TITLE
Used greedy set cover algorithm to reduce scattering batch of ids

### PR DESCRIPTION
**Pull request by ruiwang-linkedin**

### CODE REVIEW

**Comment by ymatsuda** [comment]
looks good to me.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala [hunk 1]
```diff
@@ -18,18 +18,23 @@ package network
...
 import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
+import scala.util.Random
+import scala.util.control.Breaks._
 
 /**
  * This class is intended for applications where there is a mapping from partitions -> servers able to respond to those requests. Requests are round-robined
  * between the partitions
  */
-abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartitions: Int, serveRequestsIfPartitionMissing: Boolean = true) extends PartitionedLoadBalancerFactory[PartitionedId] {
+abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartitions: Int, setLimit: Int = 0, serveRequestsIfPartitionMissing: Boolean = true) extends PartitionedLoadBalancerFactory[PartitionedId] with Logging {
```
Comment (ymatsuda) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala]: It will be helpful if you add a comment regarding the usage of the new parameter, setLimit.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala [hunk 2]
```diff
@@ -59,7 +64,75 @@ abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartition
...
+          partitionToNodeMap.get(partitionIds.head) match {
+            case None =>
+              break
+            case Some((endpoints, counter, states)) =>
+              import math._
+              val es = endpoints.size
+              counter.compareAndSet(java.lang.Integer.MAX_VALUE, 0)
+              val idx = counter.getAndIncrement % es
+              var i = idx
+              breakable {
```
Comment (jhartman) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:108]: I'm a bit nervous about performance. This will be a pretty tight loop. I don't recommend using breakable in Scala, it's known to be a bit slow.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala [hunk 3]
```diff
@@ -59,7 +64,75 @@ abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartition
           throw new InvalidClusterException("Partition %s is unavailable, cannot serve requests.".format(partition))
       }
     }
+
+    override def nodesForPartitionedIds(partitionedIds: Set[PartitionedId], capability: Option[Long]) = {
+      val partitionsMap = partitionedIds.foldLeft(Map.empty[Int, Set[PartitionedId]])
```
Comment (jhartman) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:78]: Do you think you could document the algorithm? It's not completely obvious what's going on.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala (hunk 2 above)
Comment (ruiwang-linkedin) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:108]: I've read about Scala break performance concerns. In this case, the outer break will only be broken once when the entire set is   "covered". The inner loop will be also only broken when the entire set has been covered (once). Thus, I won't be too concerned. 

I have measured the performance on both single partitionedId case vs multiple partitionedIds case, they both match the current performance numbers.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala (hunk 3 above)
Comment (ruiwang-linkedin) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:78]: Definitely. Let me add them in comments. 

In general, here is how the algorithm works.

1) calculates the partition Ids (S) for the input set of partitioned ids.
2) take the first element from the partition ids, get the nodes serving this partition, and intersect set S with each node, and find the largest match. When both capability and canServiceRequest eval to true, the node with the largest match will be put into result set. S = S - ( last intersection result)
3) Repeat 2) until S is reduced to empty.
4) The counter is for load balancing across all nodes serving the same partition. It instructs where the greedy intersection check to start, thus favoring nodes closer to where the counter points to when there is a tie.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala [hunk 4]
```diff
@@ -59,7 +64,95 @@ abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartition
...
     }
+
+    /**
+     * Use greedy set cover to minimize the nodes that serve the requested partitioned Ids
+     *
+     * @param partitionedIds
+     * @param capability
+     * @return
+     */
+    override def nodesForPartitionedIds(partitionedIds: Set[PartitionedId], capability: Option[Long]) = {
```
Comment (raymondng) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:75]: Do you have a test that validate this algorithm?

**Comment by ruiwang-linkedin** [comment]
Joe, my assumption here is canServeRequest is a relatively more expensive operation comparing with hashset intersection. Thus, by doing intersection, I would only have to check canServeRequest when there is a bigger intersection achieved.

**Comment by sungjuc** [comment]
Why the canServeRequest is so expensive? It tries to make real request to check it?

**Comment by ruiwang-linkedin** [comment]
@sungjuc , when canServeRequest is called, clientStatisticRequestStrategy will be checked together with a few other canServeRequestStrategy, it will then calculate the cluster median and the node median. And perform comparison based on the results. There are a couple of AtomicInteger access during the call as well. I am not saying this operation is expensive in the absolute sense, but it is more expensive than hash set intersection, especially when the data size are in the same order of magnitude.

File: network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala [hunk 5]
```diff
@@ -59,7 +64,95 @@ abstract class DefaultPartitionedLoadBalancerFactory[PartitionedId](numPartition
...
+                do {
+                  val ep = endpoints(i)
+
+                  // perform intersection between the set to be covered and the set the node is covering
+                  val s = ep.node.partitionIds intersect partitionIds
+
+                  // record the largest intersect
+                  if (s.size > intersect.size && ep.canServeRequests && ep.node.isCapableOf(capability))
+                  {
+                    intersect = s
```
Comment (sungjuc) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:118]: In the nodeForPartition method in DefaultLoadBalancerHelper.scala, it checks whether the endpoint can serve request and is capable then, it sets the states information. Isn't it required in this code?

-- From line 93 of DefaultLoadBalalncerHelper.scala
          if(endpoint.canServeRequests && endpoint.node.isCapableOf(capability))
            if (states(i % es).compareAndSet(true, false))
              return Some(endpoint.node)
            else if ((i - idx) >= es)

##               states(i % es).compareAndSet(false, true)

Comment (ruiwang-linkedin) [inline network/src/main/scala/com/linkedin/norbert/network/partitioned/loadbalancer/DefaultPartitionedLoadBalancerFactory.scala:118]: Good catch Joe. The states variables were used to avoid say in a case that two different capabilities were enabled on different bits, and the access patterns from clients can potentially cause some nodes always being skipped. The states were introduced was to avoid those special cases. 

While using set cover, i think the starting location in the nodes, and the greedy algorithm itself (because the requested inputs are random) can be help avoid the corner case above. That's how load was balanced before capability was introduced.

**Comment by ruiwang-linkedin** [comment]
@raymondng Migrated tests over; thanks for the reminder.
